python/pyspark/sql/dataframe.py

```diff
@@ -2046,6 +2046,40 @@ def toDF(self, *cols):
         jdf = self._jdf.toDF(self._jseq(cols))
         return DataFrame(jdf, self.sql_ctx)
 
+    @since(3.0)
+    def transform(self, func):
+        """Returns a new class:`DataFrame` according to a user-defined custom transform method.
+        This allows chaining transformations rather than using nested or temporary variables.
+
+        :param func: a user-defined custom transform function
+        This is equiavalent to a nested call:
+            actual_df = with_something(with_greeting(source_df), "crazy"))
+
+        credit to: https://medium.com/@mrpowers/chaining-custom-pyspark-transformations-4f38a8c7ae55
+
+        A more concrete example::
+        >>> sc = pyspark.SparkContext(master='local')
+        >>> spark = pyspark.sql.SparkSession(sparkContext=sc)
+        >>> from pyspark.sql.functions import lit
+        >>> def with_greeting(df):
+        ...     return df.withColumn("greeting", lit("hi"))
+        >>> def with_something(df, something):
+        ...     return df.withColumn("something", lit(something))
+        >>> data = [("jose", 1), ("li", 2), ("liz", 3)]
+        >>> source_df = spark.createDataFrame(data, ["name", "age"])
+        >>> actual_df = source_df.transform(with_greeting).transform(lambda x: with_something(x, "crazy"))
+        >>> actual_df.show()
+        +----+---+--------+---------+
+        |name|age|greeting|something|
+        +----+---+--------+---------+
+        |jose|  1|      hi|    crazy|
+        |  li|  2|      hi|    crazy|
+        | liz|  3|      hi|    crazy|
+        +----+---+--------+---------+
+        <BLANKLINE>
```

Comment:
I think we don't need this since we normalize the newline when we run doctests.

Reply:
when I run it locally it failes if I don't put the <BLANKLINE> 
I will try